### DISCUSSION TITLE
Cannot autocomplete custom variables in Pyzo

### DISCUSSION THREAD

**Discussion by wilikeit** [pyzo/pyzo#1083] [Q&A]
```python
myGlobalVar=1
def fun1():
  myLobalVar=2
  a=1
  b=a
  print(myLobal）  #Also can't pop myLobalVar

fun  #I type "fun" and it popup"fun1"
myGlobal  #connot popup "myGlobalVar"


```

Are there any special Settings required in pyzo？

**Comment by bdieterm** [comment]
The list for name autocompletion candidates is created by Pyzo like this:
* all built-ins such as `enumerate` are added
* all keywords such as **`continue`** are added
* a `dir`-like command is executed in the current shell to get a list of local and global variables ("introspection")
* [only in the editor tab but not in the shell widget:] class, function, method and member variables are obtained via a simple parser ("fictive namespace")

See method `processAutoComp` in pyzo/core/editor.py and pyzo/core/shell.py for details.

The autocompletion suggestion for "fun" to "fun1" works in your case in the editor because of the parser.
The autocompletion for `myGlobalVar` will work if that variable is present in the current shell's scope. That means, you have to evaluate the variable first by executing that code line. Then the Pyzo introspector will add that with the dir-like command to the suggestion list.
The same goes for the local variable `myLocalVar` in your function `fun1`: In that case you would have to set a breakpoint inside that function and call that function so that `myLocalVar` is in the current shell's namespace.

**Reply by wilikeit** [reply]
As you say, it does, but it doesn't seem very convenient。

Without running a script，can I just type part of the word and pop up the word like other ides. Can pyzo do that in the next release?

**Reply by bdieterm** [reply]
> Can pyzo do that in the next release?

I am not sure if you are aware that Pyzo is developed by very few unpaid volunteers in their free time.

Pyzo is a Python IDE with focus on interactivity and introspection. It is not planned to add features just because other IDEs have that. One of Pyzo's characteristics is that there are no external dependencies except a Qt wrapper.

When writing code in an interactive way, you not only have that autocompletion you wish for, but you also have features that other Python IDEs are lacking, for example key autocompletion for dictionaries.

If you wish to improve Pyzo, feel free to experiment with its source code and to contribute code.

**Reply by wilikeit** [reply]
ok, this is the advantage of pyzo, I will  try to use its powerful interactive shell.
